src/source.cpp

```diff
@@ -0,0 +1,379 @@
+    // Generation source sites from specified distribution in user input
+    for (int64_t i = 0; i < openmc_work; ++i) {
+      // initialize random number seed
+      int64_t id = openmc_total_gen*n_particles + work_index[openmc::mpi::rank] + i + 1;
+      set_particle_seed(id);
+
+      // sample external source distribution
+      source_bank[i] = sample_external_source();
+    }
+  }
+
+  // Write out initial source
+  if (openmc_write_initial_source) {
+    write_message("Writing out initial source...", 5);
+    std::string filename = path_output + "initial_source.h5";
+    hid_t file_id = file_open(filename, 'w', true);
+    write_source_bank(file_id, work_index.data(), source_bank);
+    file_close(file_id);
+  }
+}
+
+extern "C" double* rev_energy_bins_ptr();
+
+Bank sample_external_source()
+{
+  // Set the random number generator to the source stream.
+  prn_set_stream(STREAM_SOURCE);
+
+  // Determine total source strength
+  double total_strength = 0.0;
+  for (auto& s : external_sources)
+    total_strength += s.strength();
+
+  // Sample from among multiple source distributions
+  auto n_source = external_sources.size();
+  int i = 0;
+  if (n_source > 1) {
+    double xi = prn()*total_strength;
+    double c = 0.0;
+    for (i = 0; i < external_sources.size(); ++i) {
+      c += external_sources[i].strength();
+      if (xi < c) break;
+    }
+  }
+
+  // Sample source site from i-th source distribution
+  Bank site {external_sources[i].sample()};
+
+  // If running in MG, convert site % E to group
+  if (!openmc_run_CE) {
+    // Get pointer to rev_energy_bins array on Fortran side
+    double* rev_energy_bins = rev_energy_bins_ptr();
+
+    int n = num_energy_groups + 1;
+    site.E = lower_bound_index(rev_energy_bins, rev_energy_bins + n, site.E);
+    site.E = num_energy_groups - site.E;
+  }
+
+  // Set the random number generator back to the tracking stream.
+  prn_set_stream(STREAM_TRACKING);
+
+  return site;
+}
+
+//==============================================================================
+// Fortran compatibility functions
+//==============================================================================
+
+extern "C" void free_memory_source()
+{
+  external_sources.clear();
+}
+
+extern "C" double total_source_strength()
+{
+  double strength = 0.0;
+  for (const auto& s : external_sources) {
+    strength += s.strength();
+  }
+  return strength;
+}
+
+// Needed in fill_source_bank_fixedsource
+extern "C" int overall_generation();
+
+//! Fill source bank at end of generation for fixed source simulations
+extern "C" void fill_source_bank_fixedsource()
+{
+  if (path_source.empty()) {
+    // Get pointer to source bank
+    Bank* source_bank;
+    int64_t n;
+    openmc_source_bank(&source_bank, &n);
+
+    for (int64_t i = 0; i < openmc_work; ++i) {
+      // initialize random number seed
+      int64_t id = (openmc_total_gen + overall_generation())*n_particles +
+        work_index[openmc::mpi::rank] + i + 1;
```

Comment:
I've been operating under the assumption that we should indent continued lines 5 spaces like we did in Fortran, although I see the style guide doesn't mention it.  Do you have an opinion on 2 vs. 5 spaces for this indentation?

Reply:
The five space thing was a default continuation line indent in emacs for Fortran, so I would actually not propose adopting that convention going forward. Two spaces seems sufficient.

Reply:
Agreed, 2 spaces seems fine.